README.md
<p align='center'>
  <img src="https://github.com/LuisAlejandro/fb-random-post-from-feed/blob/develop/branding/banner.svg">
  <h3 align="center">Facebook random post from feed</h3>
  <p align="center">GitHub Action for posting a random article from an atom feed to a facebook page</p>
</p>

---

Current version: 0.2.0

## 🎒 Prep Work
1. Get a facebook permanent access token (explained below) using a facebook account that owns the page where you want to post messages.
2. Find the ID of the page that you want to post messages in (explained below).
3. Find the atom feed URL that contains the posts that you wish to share.

## 🖥 Workflow Usage

Configure your workflow to use `LuisAlejandro/fb-random-post-from-feed@0.2.0`,
and provide the atom feed URL you want to use as the `FEED_URL` env variable.

Provide the access token for your Facebook app as the
`FACEBOOK_ACCESS_TOKEN` env variable, set your facebook page ID as
`FACEBOOK_PAGE_ID` (as secrets). Remember, to add secrets go to your repository
`Settings` > `Secrets` > `Actions` > `New repository secret`
for each secret.

For example, create a file `.github/workflows/schedule.yml` on
a github repository with the following content:

```yml
name: Publish random post of feed hourly
on:
  schedule:
    - cron: '0 * * * *'
jobs:
  fbpost:
    runs-on: ubuntu-20.04
    steps:
      - uses: LuisAlejandro/fb-random-post-from-feed@0.2.0
        env:
          FACEBOOK_ACCESS_TOKEN: ${{ secrets.FACEBOOK_ACCESS_TOKEN }}
          FACEBOOK_PAGE_ID: ${{ secrets.FACEBOOK_PAGE_ID }}
          FEED_URL: ${{ secrets.FEED_URL }}
```

Publish your changes, activate your actions if disabled and enjoy.

## ❗ Important notes

* The action is designed to pick random posts published within the last year. You can modify this
interval by setting a `MAX_POST_AGE` env variable with an integer value in days. For example, to pick posts
from the last month set `MAX_POST_AGE: 30`.

## 👥 How to get a Facebook permanent access token

Following the instructions laid out in Facebook's [extending page tokens documentation][2] I was able to get a page access token that does not expire.

I suggest using the [Graph API Explorer][3] for all of these steps except where otherwise stated.

### 0. Create Facebook App

**If you already have an app**, skip to step 1.

1. Go to [My Apps][4].
2. Click "+ Add a New App".
3. Setup a website app.

You don't need to change its permissions or anything. You just need an app that wont go away before you're done with your access token.

### 1. Get User Short-Lived Access Token

1. Go to the [Graph API Explorer][3].
3. Select the application you want to get the access token for (in the "Facebook App" drop-down menu, not the "My Apps" menu).
4. In the "Add a Permission" drop-down, search and check "pages_manage_posts" and "pages_read_engagement".
5. Click "Generate Access Token".
6. Grant access from a Facebook account that has access to manage the target page. Note that if this user loses access the final, never-expiring access token will likely stop working.

The token that appears in the "Access Token" field is your short-lived access token.

### 2. Generate Long-Lived Access Token

Following [these instructions][5] from the Facebook docs, make a GET request to

> https://graph.facebook.com/oauth/access_token?grant_type=fb_exchange_token&client_id=**{app_id}**&client_secret=**{app_secret}**&fb_exchange_token=**{short_lived_token}**

entering in your app's ID and secret and the short-lived token generated in the previous step.

You **cannot use the Graph API Explorer**. For some reason it gets stuck on this request. I think it's because the response isn't JSON, but a query string. Since it's a GET request, you can just go to the URL in your browser.

The response should look like this:

> {"access_token":"**ABC123**","token_type":"bearer","expires_in":5183791}

"ABC123" will be your long-lived access token. You can put it into the [Access Token Debugger][7] to verify. Under "Expires" it should have something like "2 months".

### 3. Get User ID

Using the long-lived access token, make a GET request to 

> https://graph.facebook.com/me?access_token=**{long_lived_access_token}**

The `id` field is your account ID. You'll need it for the next step.

### 4. Get Permanent Page Access Token

Make a GET request to

> https://graph.facebook.com/**{account_id}**/accounts?access_token=**{long_lived_access_token}**

The JSON response should have a `data` field under which is an array of items the user has access to. Find the item for the page you want the permanent access token from. The `access_token` field should have your permanent access token. Copy it and test it in the [Access Token Debugger][7]. Under "Expires" it should say "Never".

[2]:https://developers.facebook.com/docs/facebook-login/access-tokens#extendingpagetokens
[3]:https://developers.facebook.com/tools/explorer
[4]:https://developers.facebook.com/apps/
[5]:https://developers.facebook.com/docs/facebook-login/access-tokens#extending
[6]:https://luckymarmot.com/paw
[7]:https://developers.facebook.com/tools/debug/accesstoken

## 👥 How to get a Facebook page ID

To find your Page ID:

1. From News Feed, click Pages in the left side menu.
2. Click your Page name to go to your Page.
3. Click About in the left column. If you don't see About in the left column, click See More.
4. Scroll down to find your Page ID below More Info.

## 🕵🏾 Hacking suggestions

- You can test the script locally with Docker Compose:

  * Install [Docker Community Edition](https://docs.docker.com/install/#supported-platforms) according with your operating system
  * Install [Docker Compose](https://docs.docker.com/compose/install/) according with your operating system.

      - [Linux](https://docs.docker.com/compose/install/#install-compose-on-linux-systems)
      - [Mac](https://docs.docker.com/compose/install/#install-compose-on-macos)
      - [Windows](https://docs.docker.com/compose/install/#install-compose-on-windows-desktop-systems)

  * Install a git client.
  * Fork this repo.
  * Clone your fork of the repository into your local computer.
  * Open a terminal and navigate to the newly created folder.
  * Change to the `develop` branch.

          git checkout develop

  * Create a `.env` file with the content of the environment secrets as variables, like this (with real values):

          FACEBOOK_ACCESS_TOKEN=xxxx
          FACEBOOK_PAGE_ID=xxxx
          FEED_URL=xxxx

  * Execute the following command to create the docker image (first time only):

          make image

  * You can execute the publish script with this command:

          make publish

  * Or, alternatively, open a console where you can manually execute the script and debug any errors:

          make console
          python3 entrypoint.py

  * You can stop the docker container with:
  
          make stop

  * Or, destroy it completely:
  
          make destroy
  

## Made with :heart: and :hamburger:

![Banner](https://github.com/LuisAlejandro/fb-random-post-from-feed/blob/develop/branding/author-banner.svg)

> Web [luisalejandro.org](http://luisalejandro.org/) · GitHub [@LuisAlejandro](https://github.com/LuisAlejandro) · Twitter [@LuisAlejandro](https://twitter.com/LuisAlejandro)

HISTORY.md
# Changelog


## 0.2.0 (2022-04-21)

### Changed

* Improving documentation. [Luis Alejandro Martínez Faneyth]

* Using third party library to publish posts. [Luis Alejandro Martínez Faneyth]


### Other

* Update FUNDING.yml. [Luis Alejandro]


## 0.1.1 (2020-04-16)

### Changed

* Improving readme. [Luis Alejandro Martínez Faneyth]


## 0.1.0 (2020-04-16)

### Changed

* Migrating some functions. [Luis Alejandro Martínez Faneyth]

* Adding github actions boilerplate. [Luis Alejandro Martínez Faneyth]

* Removing [REPOST] string from messages. [Luis Alejandro Martínez Faneyth]

* Refactoring for better readability. [Luis Alejandro Martínez Faneyth]

* Using 3.5 python image for docker. [Luis Alejandro Martínez Faneyth]

* Migrating from Openshift to GitLab. [Luis Alejandro Martínez Faneyth]


### Fixed

* Fixing script command. [Luis Alejandro Martínez Faneyth]


### Other

* S. [Luis Alejandro Martínez Faneyth]

* S. [Luis Alejandro Martínez Faneyth]

* S. [Luis Alejandro Martínez Faneyth]

* Creating template. [Template builder]



test.py
# -*- coding: utf-8 -*-
#
# Please refer to AUTHORS.md for a complete list of Copyright holders.
# Copyright (C) 2020-2022 Luis Alejandro Martínez Faneyth.

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import os
import time
import datetime
from urllib.request import urlopen
from html import unescape
from random import choice

from atoma import parse_rss_bytes
from pyfacebook import GraphAPI


count = 0
json_index_content = {}
access_token = os.environ.get('FACEBOOK_ACCESS_TOKEN')
page_id = os.environ.get('FACEBOOK_PAGE_ID')
feed_url = os.environ.get('FEED_URL')
max_post_age = int(os.environ.get('MAX_POST_AGE', 365))

if not feed_url:
    raise Exception('No FEED_URL provided.')

if not page_id:
    raise Exception('No FACEBOOK_PAGE_ID provided.')

graph = GraphAPI(access_token=access_token, version="13.0")

feed_data = parse_rss_bytes(urlopen(feed_url).read())
today = datetime.datetime.now()
max_age_delta = today - datetime.timedelta(days=max_post_age)
max_age_timestamp = int(max_age_delta.strftime('%Y%m%d%H%M%S'))

for post in feed_data.items:

    item_timestamp = post.pub_date.strftime('%Y%m%d%H%M%S')

    if int(item_timestamp) >= max_age_timestamp:
        json_index_content[item_timestamp] = {
            'title': post.title,
            'url': post.guid,
            'date': post.pub_date
        }

random_post_id = choice(list(json_index_content.keys()))
random_post_title = json_index_content[random_post_id]['title']
status_link = '{0}#{1}'.format(
    json_index_content[random_post_id]['url'],
    today.strftime('%Y%m%d%H%M%S'))
data = {'message': unescape(random_post_title), 'link': status_link}

fb = graph.post_object(object_id=page_id,
                       connection='feed',
                       data=data)
time.sleep(10)
graph.delete_object(object_id=fb['id'])


utils.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
from datetime import datetime
from urllib.parse import quote
from html.parser import HTMLParser

if not sys.version_info < (3,):
    unicode = str
    basestring = str


def u(u_string):
    """
    Convert a string to unicode working on both python 2 and 3.

    :param u_string: a string to convert to unicode.

    .. versionadded:: 0.1.5
    """
    if isinstance(u_string, unicode):
        return u_string
    return u_string.decode('utf-8')


def s(s_string):
    """
    Convert a byte stream to string working on both python 2 and 3.

    :param s_string: a byte stream to convert to string.

    .. versionadded:: 0.1.5
    """
    if isinstance(s_string, bytes):
        return s_string
    return s_string.encode('utf-8')


def html_unescape(_string):
    return HTMLParser().unescape(_string)


def escape(_string):
    return quote(u(_string).encode(), safe='~')


def filter_json_index_by_year(json_index_content):
    json_index_filtered = {}
    current_year = int(datetime.now().strftime('%Y'))
    for pid, data in json_index_content.items():
        post_year = int(data['date'].strftime('%Y'))
        if post_year >= (current_year - 2):
            json_index_filtered[pid] = data
    return json_index_filtered


random_post_twitter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import json
from random import getrandbits, choice
import base64
import hmac
import hashlib
from urllib.request import urlopen, Request
from urllib.parse import urlencode
from datetime import datetime

from utils import utfize, html_unescape, escape, filter_json_index_by_year

twitter_api_version = '1.0'
twitter_api_method = 'HMAC-SHA1'
twitter_api_end = 'https://api.twitter.com/1.1/statuses/update.json'
twitter_consumer_key = 'EfRsYNSXgd62lFtuP1RahQ'
twitter_consumer_secret = '5EAk7IYlUQe3GX00bONoHx0fDJbuVUYMMDfEbbIsuY'
twitter_oauth_token = '320430429-896na1d2phjByw67KI3jr1poGtW5V2jJcZp9Zp1x'
twitter_oauth_secret = 'pAw4XsOjqBBOqaeLw7TtnwHQrY12dLdktRZ9GQyMyGwoA'
json_index_url = 'http://huntingbears.com.ve/static/json/index.json'

print('Starting publication of random post to Twitter')

current_timestamp = int(datetime.now().strftime('%Y%m%d%H%M%S'))
current_hour = int(datetime.now().strftime('%H'))

if current_hour not in [2, 6, 9, 14, 17, 21]:
    print('Script wasnt called in a recommended hour. Aborting.')
    sys.exit(0)

json_index_content = json.loads(str(urlopen(json_index_url).read(), 'utf-8'))
json_index_filtered = filter_json_index_by_year(json_index_content)

if not json_index_filtered:
    print('There are no posts to publish. Aborting.')
    sys.exit(0)

random_post_id = choice(list(json_index_filtered.keys()))
random_post_title = json_index_content[random_post_id]['title']
random_post_title = utfize(html_unescape(random_post_title))
random_post_url = utfize('{0}#{1}'.format(
    json_index_filtered[random_post_id]['url'],
    current_timestamp))

twitter_status = '{0} {1}'.format(random_post_title, random_post_url)

signature_timestamp = datetime.now().strftime('%s')
signature_once = str(getrandbits(64))
signature_basestr_auth = {}
signature_basestr_auth['status'] = twitter_status
signature_basestr_auth['oauth_consumer_key'] = twitter_consumer_key
signature_basestr_auth['oauth_nonce'] = signature_once
signature_basestr_auth['oauth_signature_method'] = twitter_api_method
signature_basestr_auth['oauth_timestamp'] = signature_timestamp
signature_basestr_auth['oauth_token'] = twitter_oauth_token
signature_basestr_auth['oauth_version'] = twitter_api_version

signature_basestr_auth = ['{0}={1}'.format(escape(k), escape(v))
                          for k, v in sorted(signature_basestr_auth.items())]
signature_basestr_auth = '&'.join(signature_basestr_auth)

signature_basestr = 'POST&{0}&{1}'.format(escape(twitter_api_end),
                                          escape(signature_basestr_auth))
signature_key = '{0}&{1}'.format(twitter_consumer_secret,
                                 escape(twitter_oauth_secret))

oauth_hmac_hash = hmac.new(signature_key.encode(), signature_basestr.encode(),
                           hashlib.sha1)
oauth_signature = base64.b64encode(oauth_hmac_hash.digest()).decode()

oauth_header = {}
oauth_header['oauth_consumer_key'] = twitter_consumer_key
oauth_header['oauth_nonce'] = signature_once
oauth_header['oauth_signature'] = oauth_signature
oauth_header['oauth_signature_method'] = twitter_api_method
oauth_header['oauth_timestamp'] = signature_timestamp
oauth_header['oauth_token'] = twitter_oauth_token
oauth_header['oauth_version'] = twitter_api_version

oauth_header = ['{0}="{1}"'.format(escape(k), escape(v))
                for k, v in sorted(oauth_header.items())]
oauth_header = ', '.join(oauth_header)
oauth_header = 'OAuth {0}'.format(oauth_header)

http_headers = {'Authorization': oauth_header,
                'Content-Type': 'application/x-www-form-urlencoded'}
http_request = Request(url=twitter_api_end,
                       data=urlencode({'status': twitter_status}).encode(),
                       headers=http_headers)

count = 0
while count < 6:
    try:
        result = json.loads(str(urlopen(http_request).read(), 'utf-8'))
    except Exception as e:
        print('There was an error publishing: {0}'.format(e))
        count += 1
        continue

    if 'errors' in result:
        count += 1
        continue

    print('Successfully published!: {0}'.format(random_post_url))
    break


random_post_facebook.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# http://stackoverflow.com/questions/17197970/facebook-permanent-page-access-token

import sys
import json
from random import choice
from urllib.request import urlopen, Request
from urllib.parse import urlencode
from datetime import datetime

from utils import utfize, html_unescape, filter_json_index_by_year

facebook_api_end = 'https://graph.facebook.com/156686441067299/feed'
facebook_access_token = ('EAAF0cGsgXS8BAMniHEKNzK5KPeaWcevyCbu8lwQH0Y'
                         'FWsLj5VWyvcEyknMh1IYmIxXpZCbJiZC6Slfg3rZBzm'
                         'R8ZCRpo52sHx7hCWVUXM4Jw3w3aOfU9ANSRZBLcXvFm'
                         'uEEBQSfHno51gvTAS7d1f0QtrYTxzqZBrSuGwr0q3cwAZDZD')
json_index_url = 'http://huntingbears.com.ve/static/json/index.json'

print('Starting publication of random post to Facebook')

current_timestamp = int(datetime.now().strftime('%Y%m%d%H%M%S'))
current_hour = int(datetime.now().strftime('%H'))

if current_hour not in [9, 13, 15, 22]:
    print('Script wasnt called in a recommended hour. Aborting.')
    sys.exit(0)

json_index_content = json.loads(str(urlopen(json_index_url).read(), 'utf-8'))
json_index_filtered = filter_json_index_by_year(json_index_content)

if not json_index_filtered:
    print('There are no posts to publish. Aborting.')
    sys.exit(0)

random_post_id = choice(list(json_index_filtered.keys()))
random_post_title = json_index_filtered[random_post_id]['title']
random_post_title = utfize(html_unescape(random_post_title))
random_post_url = utfize('{0}#{1}'.format(
    json_index_filtered[random_post_id]['url'],
    current_timestamp))

facebook_api_data = {'message': random_post_title,
                     'link': random_post_url,
                     'access_token': facebook_access_token}

http_request = Request(url=facebook_api_end, method='POST',
                       data=urlencode(facebook_api_data).encode())

count = 0
while count < 6:
    try:
        result = json.loads(str(urlopen(http_request).read(), 'utf-8'))
    except Exception as e:
        print('There was an error publishing: {0}'.format(e))
        count += 1
        continue

    if 'error' in result:
        count += 1
        continue

    print('Successfully published!: {0}'.format(random_post_url))
    break


entrypoint.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import json
from random import choice
from urllib.request import urlopen, Request
from urllib.parse import urlencode
from datetime import datetime

from atoma import parse_rss_bytes

from utils import utfize, html_unescape, filter_json_index_by_year


json_index_content = {}
facebook_page_id = os.environ.get('FACEBOOK_PAGE_ID')
facebook_access_token = os.environ.get('FACEBOOK_ACCESS_TOKEN')
facebook_api_end = 'https://graph.facebook.com/{0}/feed'.format(facebook_page_id)
feed_url = os.environ.get('FEED_URL')
feed_data = parse_rss_bytes(urlopen(feed_url).read())

current_timestamp = int(datetime.now().strftime('%Y%m%d%H%M%S'))
current_hour = int(datetime.now().strftime('%H'))

if current_hour not in [9, 13, 15, 22]:
    print('Script wasnt called in a recommended hour. Aborting.')
    sys.exit(0)

for post in feed_data.items:
    post_timestamp = post.pub_date.strftime('%Y%m%d%H%M%S')
    json_index_content[post_timestamp] = {
        'title': post.title,
        'url': post.guid,
        'date': post.pub_date
    }

json_index_filtered = filter_json_index_by_year(json_index_content)

if not json_index_filtered:
    print('There are no posts to publish. Aborting.')
    sys.exit(0)

random_post_id = choice(list(json_index_filtered.keys()))
random_post_title = json_index_filtered[random_post_id]['title']
random_post_title = utfize(html_unescape(random_post_title))
random_post_url = utfize('{0}#{1}'.format(
    json_index_filtered[random_post_id]['url'],
    current_timestamp))

facebook_api_data = {'message': random_post_title,
                     'link': random_post_url,
                     'access_token': facebook_access_token}

http_request = Request(url=facebook_api_end, method='POST',
                       data=urlencode(facebook_api_data).encode())

count = 0
while count < 6:
    try:
        result = json.loads(str(urlopen(http_request).read(), 'utf-8'))
    except Exception as e:
        print('There was an error publishing: {0}'.format(e))
        count += 1
        continue

    if 'error' in result:
        count += 1
        continue

    print('Successfully published!: {0}'.format(random_post_url))
    break
